presto-main/src/main/java/com/facebook/presto/operator/PartitionedOutputOperator.java

```diff
@@ -399,6 +476,48 @@ public void partitionPage(Page page)
             flush(false);
         }
 
+        void ariaPartitionPage(Page page)
+        {
+            calculateRowSizes(page);
+            populateRowNumbersForEachPartition(page);
+
+            for (int i = 0; i < partitionData.length; i++) {
+                partitionData[i].appendRows(page, outputBuffer);
+            }
+        }
+
+        private void calculateRowSizes(Page page)
+        {
+            if (rowSizes == null || rowSizes.length < page.getPositionCount()) {
+                rowSizes = new int[(int) (page.getPositionCount() * 1.2)];
+            }
+            Arrays.fill(rowSizes, 0);
+            for (int i = 0; i < page.getChannelCount(); i++) {
+                page.getBlock(i).accumulateSizesInBytes(rowSizes);
+            }
+        }
+
+        private void populateRowNumbersForEachPartition(Page page)
+        {
+            for (PartitionData target : partitionData) {
+                target.positionCount = 0;
```

Comment:
Add reset method to PartitionData:

```
        public void reset()
        {
            positionCount = 0;
        }

            for (PartitionData partition : partitionData) {
                partition.reset();
            }
```

Reply:
I think reset() is too generic and reader might think this also resets the buffers. I renamed it to resetPositionCount() for now, and we can revisit the naming later.